src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs

```diff
@@ -0,0 +1,428 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Common;
+using Common.Log;
+using Lykke.Bitcoin.Api.Client.AutoGenerated.Models;
+using Lykke.Bitcoin.Api.Client.BitcoinApi;
+using Lykke.Bitcoin.Api.Client.BitcoinApi.Models;
+using Lykke.Service.PayInternal.Core;
+using Lykke.Service.PayInternal.Core.Domain.Transfer;
+using Lykke.Service.PayInternal.Core.Domain.Wallet;
+using Lykke.Service.PayInternal.Core.Services;
+using Lykke.Service.PayInternal.Services.Domain;
+using NBitcoin;
+using Newtonsoft.Json;
+
+
+namespace Lykke.Service.PayInternal.Services
+{
+    public class BtcTransferRequestService : ITransferRequestService
+    {
+
+        private readonly ITransferRepository _transferRepository;
+        private readonly IWalletRepository _walletRepository;
+        private readonly IBitcoinApiClient _bitcointApiClient;
+        private readonly ITransferRequestPublisher _transferRequestPublisher;
+        private readonly ILog _log;
+
+
+        #region .ctors
+        public BtcTransferRequestService(
+            ITransferRepository transferRepository,
+            IWalletRepository walletRepository,
+            IBitcoinApiClient bitcointApiClient,
+            ITransferRequestPublisher transferRequestPublisher,
+            ILog log)
+        {
+            _transferRepository = transferRepository ?? throw new ArgumentNullException(nameof(transferRepository));
+            _walletRepository = walletRepository ?? throw new ArgumentNullException(nameof(walletRepository));
+            _bitcointApiClient = bitcointApiClient ?? throw new ArgumentNullException(nameof(bitcointApiClient));
+            _transferRequestPublisher = transferRequestPublisher ?? throw new ArgumentNullException(nameof(transferRequestPublisher));
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+        }
+        #endregion
+
+        #region public
+
+        /// <summary>
+        /// Tries to execute a crosswise transfer request from many sources to many destinations.
+        /// </summary>
+        /// <param name="transferRequest">The transfer request.</param>
+        /// <returns>The updated transfer request OR transfer request with error.</returns>
+        public async Task<ITransferRequest> CreateTransferCrosswiseAsync(ITransferRequest transferRequest)
+        {
+            // We need an independent copy of the request here, for some methods in current implementation change the inner state of the given object.
+            // For instance, the BtcTransferRequest class' constructor reuses the transaction list it gets.
+            var transferRequestUpdated = transferRequest.DeepCopy();
```

Comment:
Do you share the same object among differenct services? Didn't find it. The idea of copying the object seems strange. Anyway, why do you need to update domain request object? If it is request, it is supposed to bring information about request into domain service.

Reply:
I mean, the payload of request is to deliver information but not to serve as an object with some status

Reply:
And again, here we need our Core Comment-1.

Additionally, it this part of code we do the following:
- obtain (by ref) some object representing the transfer;
- try to execute it's transaction modifying the state of our transfer (obtained by ref);
- and then return to the caller our updated transfer request -or- "error transfer request".

So, we need to separate the input object from the returned one.